File: frontend/src/types/geo.js
/**
 * @fileoverview Geographic type definitions for RASED
 * These are JSDoc typedefs mirroring the TypeScript types in the architecture doc.
 */

/**
 * @typedef {Object} CommuneMapFeature
 * @property {string} commune_code
 * @property {string} commune_name
 * @property {string} wilaya_code
 * @property {'0'|'I'|'IIa'|'IIb'|'III'} zone_sismique
 * @property {number} lat
 * @property {number} lon
 * @property {number} total_exposure
 * @property {number} policy_count
 * @property {number} avg_risk_score
 * @property {number} net_retention
 * @property {number} hotspot_score
 */

/**
 * @typedef {Object} PortfolioKPIs
 * @property {number} total_exposure
 * @property {number} total_policies
 * @property {number} net_retention
 * @property {{ zone: string, exposure: number, policy_count: number, pct: number }[]} by_zone
 * @property {HotspotData} top_hotspot
 */

/**
 * @typedef {Object} HotspotData
 * @property {string} commune_name
 * @property {string} wilaya_code
 * @property {string} zone_sismique
 * @property {number} hotspot_score
 * @property {number} total_exposure
 * @property {number} policy_count
 */

/**
 * @typedef {Object} WilayaBasic
 * @property {string} code
 * @property {string} name_fr
 * @property {string} zone_sismique
 */

/**
 * @typedef {Object} CommuneBasic
 * @property {number} id
 * @property {string} commune_code
 * @property {string} commune_name
 * @property {string} zone_sismique
 */

/**
 * @typedef {Object} PremiumAdequacyRow
 * @property {string} zone_sismique
 * @property {string} type_risque
 * @property {number} adequate_rate
 * @property {number} avg_actual_rate
 * @property {number} gap_pct
 */

/** @typedef {'risk'|'exposure'|'score'|'simulation'} MapLayer */

export const ZONE_COLORS = {
  '0':   '#22c55e',   // green
  'I':   '#84cc16',   // lime
  'IIa': '#eab308',   // yellow
  'IIb': '#f97316',   // orange
  'III': '#dc2626',   // red
}

export const ZONE_WEIGHTS = {
  '0': 1, 'I': 2, 'IIa': 3, 'IIb': 4, 'III': 5,
}

File: backend/app/schemas/geo.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel

MapLayer = Literal["risk", "exposure", "score", "simulation"]


class WilayaBasic(BaseModel):
    code: str
    name: str
    commune_count: int | None = None


class CommuneBasic(BaseModel):
    id: int | None = None
    code: str | None = None
    name: str
    wilaya_code: str
    wilaya_name: str
    zone_sismique: str
    zone_num: int | None = None
    zone_source: str | None = None
    lat: Decimal | None = None
    lon: Decimal | None = None
    coordinate_source: str | None = None
    has_coordinates: bool = False


class ZoneLookupResponse(BaseModel):
    wilaya_code: str
    wilaya_name: str | None = None
    commune_code: str | None = None
    commune: str
    zone: str
    zone_num: int | None = None
    zone_source: str | None = None
    description: str
    lat: Decimal | None = None
    lon: Decimal | None = None
    coordinate_source: str | None = None
    has_coordinates: bool = False


class CommuneMapFeature(BaseModel):
    commune_code: str | None
    commune_name: str
    wilaya_code: str
    wilaya_name: str
    zone_sismique: str
    zone_source: str | None = None
    lat: Decimal | None = None
    lon: Decimal | None = None
    coordinate_source: str | None = None
    has_coordinates: bool = False
    total_exposure: Decimal
    policy_count: int
    avg_risk_score: Decimal
    net_retention: Decimal
    hotspot_score: Decimal
    layer_value: Decimal


class MapDataResponse(BaseModel):
    features: list[CommuneMapFeature]
    last_updated: datetime | None


class HotspotData(BaseModel):
    rank: int
    wilaya_code: str
    wilaya_name: str
    commune_code: str | None
    commune_name: str
    zone_sismique: str
    zone_source: str | None = None
    lat: Decimal | None = None
    lon: Decimal | None = None
    coordinate_source: str | None = None
    has_coordinates: bool = False
    total_exposure: Decimal
    policy_count: int
    hotspot_score: Decimal


class ZoneBreakdown(BaseModel):
    zone: str
    exposure: Decimal
    policy_count: int
    pct: Decimal


class PortfolioKPIs(BaseModel):
    total_exposure: Decimal
    total_policies: int
    net_retention: Decimal
    by_zone: list[ZoneBreakdown]
    top_hotspot: HotspotData | None


class PremiumAdequacyRow(BaseModel):
    zone: str
    type_risque: str
    adequate_rate: Decimal
    observed_rate: Decimal
    premium_gap_pct: Decimal
    policy_count: int
    total_exposure: Decimal

